### PR TITLE
Integration specs

### PR DESCRIPTION
**Pull request by Ogaga01**
# Hello Reviewer 👋 

In this project milestone, I implemented the following requirements:
#### n+1
- Make sure that the N+1 problem is solved when fetching all posts and their comments for a user.
    - Do not use the bullet gem in this exercise - instead, analyze the logs as shown in this [article]
    - **In your PR description, include the screenshots of your app's logs before and after the fix**
    - This is the corresponding wireframe view:blog all user posts

#### Integration specs
- Use Capybara to write integration tests for each view in your project.
- Think about the flow of things and group the tests of the page in a way that makes sense. For example, one test would check for the things on the page and another for clicking links
Screenshot (7)
Screenshot (8)
Screenshot (9)
Screenshot (10)
.

- User index page:
  - I can see the username of all other users.
  - I can see the profile picture for each user.
  - I can see the number of posts each user has written.
  - When I click on a user, I am redirected to that user's show page.

- user show page:
  - I can see the user's profile picture.
  - I can see the user's username.
  - I can see the number of posts the user has written.
  - I can see the user's bio.
  - I can see the user's first 3 posts.
  - I can see a button that lets me view all of a user's posts.
  - When I click a user's post, it redirects me to that post's show page.
  - When I click to see all posts, it redirects me to the user's post's index page.

- User post index page:
  - I can see the user's profile picture.
  - I can see the user's username.
  - I can see the number of posts the user has written.
  - I can see a post's title.
  - I can see some of the post's body.
  - I can see the first comments on a post.
  - I can see how many comments a post has.
  - I can see how many likes a post has.
  - I can see a section for pagination if there are more posts than fit on the view.
  - When I click on a post, it redirects me to that post's show page.

- Post-show page:
  - I can see the post's title.
  - I can see who wrote the post.
  - I can see how many comments it has.
  - I can see how many likes it has.
  - I can see the post body.
  - I can see the username of each comment.
  - I can see the comment each comment left.

**Include integration specs for any other features you may have implemented!**

### CODE REVIEW

**Review by BregornOriginal — COMMENTED** [review]
Hi @Ogaga01  

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request

***You have submitted a project, that doesn´t follow a `pair-programming` activity. You should ask your mate to help you with the test and do some commits to the project to be validated***

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

File: spec/features/user_spec.rb [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+      @user_third = User.create(name: 'Ogaga', photo: 'tom.png', bio: 'A good guy', post_counter: 2)
+    end
+
+    it 'Should display all user names' do
+      visit '/'
+      expect(page).to have_content('Tom')
+      expect(page).to have_content('Lilly')
+      expect(page).to have_content('Ogaga')
+    end
+
```
Comment (Davidosky007) [inline spec/features/user_spec.rb:17]: * Kindly add a test to cover this `I can see the profile picture for each user.` as its missing

File: spec/features/user_spec.rb [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    it 'Should display the user name' do
+      visit user_path(@user_first.id)
+      expect(page).to have_content('Tom')
+    end
+
+    it 'Should display the total post number of 3' do
+      visit user_path(@user_first.id)
+      expect(page).to have_content('Number of posts: 3')
+    end
```
Comment (Davidosky007) [inline spec/features/user_spec.rb]: * This test fails, Kindly fix it 
image
## Hints 💭 you can solve this test by targeting the actual 3posts you created for example 👇 
image

File: spec/features/user_spec.rb [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+      visit user_path(@user_first.id)
+      expect(page).to have_content('This is my first post')
+      expect(page).to have_content('This is my second post')
+      expect(page).to have_content('This is my third post')
+    end
+
+    it 'Should display a button with "See all posts" text' do
+      visit user_path(@user_first.id)
+      expect(page).to have_content('see all post')
+    end
```
Comment (Davidosky007) [inline spec/features/user_spec.rb:86]: * Kindly write a test to cover this scenario `When I click a user's post, it redirects me to that post's show page` as its missing as its a [requirement](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md#:~:text=When%20I%20click%20a%20user%27s%20post%2C%20it%20redirects%20me%20to%20that%20post%27s%20show%20page)

File: spec/features/post_spec.rb [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+      @like = Like.new(author_id: @user_first.id, post_id: @post_first.id)
+      @like.author = @user_first
+      @like.post = @post_first
+      @like.save
+      @like.update_posts_likes_counter
+    end
+
+    it 'Should set the link of the picture to the img src' do
+      visit user_posts_path(@user_first.id)
```
Comment (Davidosky007) [inline spec/features/post_spec.rb]: * Kindly update your tests as some scenarios are missing below 👇 is a screenshot of the missing tests required 
image

File: spec/features/post_spec.rb [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    end
+
+    it 'Should display the total post likes number of 1' do
+      visit user_post_path(@user_first.id, @post_first.id)
+      expect(page).to have_content('Likes: 1')
+    end
+
+    it 'Should display the post body' do
+      visit user_post_path(@user_first.id, @post_first.id)
+      expect(page).to have_content('This is my first post')
```
Comment (Davidosky007) [inline spec/features/post_spec.rb]: * These tests are failing , Kindly fix them 
image

**Review by Davidosky007 — CHANGES_REQUESTED** [review]
Hi @Ogaga01 @BregornOriginal 👋, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
## Highlights
* N+1 problems solved 👍 
* Well written PR 👍 
* No linter errors 👍 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @davidsoky007 in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/features/user_spec.rb [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+      visit user_path(@user_first.id)
+      click_link 'see all post'
+      expect(page).to have_current_path(user_posts_path(@user_first.id))
+    end
+
+    it "Should redirect me to that post's show page" do
+      visit user_path(@user_first.id)
+      click_on 'Posts'
+      expect(page).to have_current_path(user_post_path(@user_first.id))
+    end
```
Comment (codepantha) [inline spec/features/user_spec.rb]: - [x] This post is failing. Kindly fix it ✔️ 
image

File: spec/features/post_spec.rb [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    it 'I can see a section for pagination if there are more posts than fit on the view' do
+      visit user_posts_path(@user_first.id)
+      expect(page).to have_content('Tom')
+    end
+
+    it "Should redirect me to that post's show page" do
+      visit user_path(@user_first.id)
+      click_on 'Posts'
+      expect(page).to have_current_path(user_post_path(@user_first.id))
+    end
```
Comment (codepantha) [inline spec/features/post_spec.rb]: - [x] This test is failing. Please try to make it pass ✔️ 
image

File: spec/features/post_spec.rb [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+    it 'I can see how many likes a post has' do
+      visit user_posts_path(@user_first.id)
+      expect(@post_first.likes_counter).to eq 3
+    end
+
+    it 'I can see a section for pagination if there are more posts than fit on the view' do
+      visit user_posts_path(@user_first.id)
+      expect(page).to have_content('Tom')
+    end
```
Comment (codepantha) [inline spec/features/post_spec.rb]: - [x] This test is not testing what it should be. Please write a test that actually asserts that a section for pagination is seen
image

**Review by codepantha — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED ♻️ 

## Hi @Ogaga01 👋, 

Great work on making the changes requested by the previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.



## Highlights 🎯 
✔️ Descriptive PR
✔️ Solved n+1 problem 📌 
✔️ Good commit history
✔️ No linter errors

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/features/post_spec.rb [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+    it 'I can see how many likes a post has' do
+      visit user_posts_path(@user_first.id)
+      expect(@post_first.likes_counter).to eq 3
+    end
+
+    it 'I can see a section for pagination if there are more posts than fit on the view' do
+      visit user_posts_path(@user_first.id)
+      expect(@user_first.post_counter).to eq 1
+    end
```
Comment (Gambit142) [inline spec/features/post_spec.rb]: - [x] The test for pagination is not correct as highlighted by the previous reviewer. On the `Post#Index` page, there ought to be a button to implement the _**Pagination feature**_. Even though it wasn't a hard requirement, the button ought to exist as a placeholder. It is the presence of the button on the `Post#Index` page that you ought to test for. Kindly implement this change. 👍 😎

**Review by Gambit142 — CHANGES_REQUESTED** [review]
Hi @Ogaga01 @jodavid444 👋, 

Great work on making the changes requested by a previous reviewer 👏 

## Highlights
- All linters checks passed ✔️ 
- Use of capybara gem to test for the project's features ✔️ 
- Wrote tests for _**Show**_ and _**Index**_ pages for `Post` and `User` controllers ✔️ 

You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

 ## Required Changes ♻️
_Kindly check the comments under the review._

## Optional suggestions

N/A

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will improve your application._


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

#### Please, remember to tag me in your question so I can receive the notification. You can tag me by typing @Gambit142

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by codepantha — APPROVED** [review]
# STATUS: APPROVED 🟢 ✔️ 

## Hi @Ogaga01 👋, 

Great job implementing the required changes 👏 
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**